### DISCUSSION TITLE
Livewire date casting and validation problem?

### DISCUSSION THREAD

**Discussion by limeframe** [livewire/livewire#5632] [1. Help]
First of all, i think that is a bug but i will try :)

I am trying to implement a date validation on a livewire component, and as i understand it appears that date casting happens before validation resulting with errors.

**In the model**
```php
protected $casts = [
    'start_date' => 'date',
];

protected $appends = ['start_date_new'];

public function getStartDateNewAttribute()
{
    return Carbon::parse($this->start_date)->format('d/m/Y');
}

public function setStartDateNewAttribute($value)
{
     dd('this fired before validation takes place!!!!');
     $this->start_date = Carbon::createFromFormat('d/m/Y', $value)->format('d-m-Y');
}
```

**In the livewire validation rules**
```php
protected $rules = [
    'editing.start_date_new'      => 'required',
];
```

Thanks

**Comment by limeframe** [comment]
Anyone an idea?
Thanks

**Comment by joshhanley** [comment]
@limeframe honestly, model binding to attributes that have casts like that can cause issues and is near impossible to validate (as you've found). It would be better to just bind your date input to a property on the component, that you can then run validation against, and then convert to a carbon instance once you are persisting the changes to your model.

Hope this helps!

**Comment by chrisdicarlo** [comment]
@joshhanley It might be worth adding a caveat to the docs to make sure people know that the validation system does not work as expected when binding to Eloquent properties that have casts.  I just spent about an hour troubleshooting this issue with a date cast on a model.  It's not a big deal having discrete properties (though the additional boilerplate is a bit of pain when saving, etc) but when reading the docs, the impression is that it **should just work** when binding to Eloquent model properties.

Having some validation rules set on the Model and then others as discrete property rules (if they are cast on the model) feels dirty, IMHO.

After debugging, I get why it's happening - mutator firing before the validation.  Is there a way to selectively defer a mutator until after validation occurs??

I kind of feel now that for the sake of future maintenance, I can't really trust the Eloquent property binding but that also makes one of the truly beautiful things about working with Livewire a PITA if you have large forms with a huge number of properties.

I didn't see anything mentioning this edge case in the docs; if it's there, I apologize!

Cheers,
Chris
